### DISCUSSION TITLE
Keypad motion sensor

### DISCUSSION THREAD

**Discussion by exr4361** [tsightler/ring-mqtt#526] [Feature Requests]
Amazing addon! I appreciate all your hard work on this. Would it be possible to expose the motion sensor on the Keypad V2 and configure the motion sensor timeout to setup automations with that motion sensor. Looking forward to your response. Thanks!

**Comment by tsightler** [comment]
This has been asked before but I don't think it's possible.  I poked around at it a little bit, but I don't see any event data coming back from the keypad when motion turns the keypad backlight on.  It appears to me that this function is handled completely within the device hardware.  Pretty much, if there's no indication in the Ring app, then it probably isn't possible with ring-mqtt either.

**Comment by GizmoGoody** [comment]
Vote Here: https://community.ring.com/t/keypad-motion-detector/30383

**Comment by tsightler** [comment]
Recently dug into this a bit deeper and see that, while this data isn't presented in the device data structure in ring-client-api, there are actually events available and you can detect, for example, a proximity motion event vs a button-press based on the impulse type in the raw data update.  This means that it is at least technically possible to use the proximity sensor to detect and report motion by monitoring for events with impulseType "keypad.motion".

Based on this, I'm going to add this to the "to-do" list and see if I can figure out a way to implement this practically.

**Reply by tsightler** [reply]
Dev branch has a simple implementation that seems to work.

**Reply by GizmoGoody** [reply]
Nice! I'll have to wait until the master branch is updated since I'm running HAOS.

**Reply by tsightler** [reply]
You can use the branch feature to test it out, just change branch config option from "addon" to "dev" and restart the addon and it will download the lastest dev branch during the next startup.  Just switch it back to "addon" to revert to the built-in code.

**Reply by GizmoGoody** [reply]
Thanks. I did and it seems to be working well so far. I have it in an automation that only fires once a day, but I'll let you know if I experience any issues. Thanks again for adding it!

**Reply by GizmoGoody** [reply]
I might have spoken too soon or this could be a coincidence and something else is going on. Just starting today, when one of the zones (a door) that I have configured through a retrofit kit is opened, all Ring MQTT entities go unavailable for ~2 minutes, then they all come back. The other three doors configured this way do not cause this and the "trouble" zone responds as expected in the Ring app. I'll continue to monitor, but any chance this could be related to changes on the dev branch?

**Reply by tsightler** [reply]
Because the code in question has to monitor for all device updates it might be causing some type of exception when a specific message comes through, impossible to really know without logs of the crash event.

**Reply by GizmoGoody** [reply]
Gotcha. Which logs (and where in the UI) should I enable and monitor?

**Reply by tsightler** [reply]
The logs should be readily available in the UI on the logs tab of the addon, however, you may have a difficult time catching the error since the UI only shows the last 50 lines or so, although maybe since you can trigger it by opening a door, you can catch it.  I'm not aware of any way to get the full logs via the UI.  In "Settings...System...Logs" you can select the ring-mqtt addon and there is a button to "Download full logs", but, as far as I can tell, it still only downloads the logs shown in the UI, so not really the full logs.

The only way I know to get the full logs is the procedure I've documented here:

https://github.com/tsightler/ring-mqtt-ha-addon/blob/main/GET-LOGS.md

Unfortunately, I am not going to have much time over the next couple of weeks as I'm traveling, so I most likely won't be able to look at it until I get back.

**Reply by tsightler** [reply]
Sorry for the delay, I've been out of town for a few weeks.  I never received any logs from you, but I think I finally managed to track down the crashing issues and I believe it is fixed in the lastest dev code.  I hope to push this out in the next few days if it seems stable after some additional testing.

**Reply by GizmoGoody** [reply]
Sorry, I'm out of town for a couple weeks myself, so I had to put it back to the stable release before leaving. I'll try the latest when I get home and let you know. Thank you.

**Reply by tsightler** [reply]
No problem.  I went ahead and pushed 5.2.0 as I haven't seen any more crashes.  Hopefully it's clean.

**Reply by alvinchen1** [reply]
Hi!  Looks like it is working. Thanks so much! The entity appeared as soon as the update was applied.  However, it didn't start actually functioning until I changed the Keypad - LED Settings to either Keep Mode Button Lit or Motion Turns on Lights.
The option I had all my keypads set to was Button Press Turns on Lights.

![Screenshot_20230314-145018](https://user-images.githubusercontent.com/26409535/225107981-944221d7-5a2c-489f-aa84-13f8dd143bcf.png)
(Screenshot of Setting in Android Ring App v3.57.1)

Seems like if 'Button Press Turn on Lights' is set, motion sensor does not work.

Would suggest on the Wiki, adding under 

- Keypad
   - Motion Detection (LED Setting set to either 'Keep Mode Button Lit' or 'Motion Turns on Lights')
